Gutenboarding: Separate DomainPicker and onboarding concerns

sirreal opened the pull request:
- Reorganize the domain picker internals, the selection happens closer to where it's required.
- Isolate the DomainPicker from onboarding store.
- The contents of the DomainPicker button is now passed as `children`.
- Only show the domain picker if we have a domain already selected or a site title.

## Testing
- `/gutenboarding` domains picker works well
- Domains picker appears when appropriate (have site title) and responds to changes to site title appropriately
- Domains picker searches correctly when a term is entered.

ockham commented on client/landing/gutenboarding/stores/onboard/types.ts:
```diff
@@ -38,3 +38,5 @@ export interface Vertical {
 	synonyms: string[];
 }
+
+export type DomainName = string;
```
Shouldn't define here IMO, since we're already defining it in the domain suggestions store, and the definitions could collide. I'd suggest just changing the import paths for the `DomainName` type in this store's reducer and actions file (which I'd messed up in the first place :roll_eyes: )

ockham replied:
(Per a video call, let's remove this from this PR and tackle separately.)

ockham commented:
Collecting some (edge) cases:

Enter something into the site title field, then delete again. DomainPickerButton continues to show the free domain for that site title.

ockham commented:
I also _think_ we should never show 'Choose a domain' but hide the picker as long as we don't have any site title information.

sirreal commented on client/landing/gutenboarding/components/header/index.tsx:
```diff
@@ -49,7 +86,7 @@ const Header: FunctionComponent< Props > = ( {
 						{ siteTitle ? siteTitle : NO__( 'Create your site' ) }
 					</div>
-					<DomainPickerButton />
+					<DomainPickerButton>{ domainText ?? NO__( 'Choose a domain' ) }</DomainPickerButton>
```
Don't show if we don't have enough information.

```suggestion
					{ domainText && <DomainPickerButton>{ domainText }</DomainPickerButton> }
```

ockham commented:
I seem to get different free WP.com subdomain suggestions for Roasting Beans in the Domain Picker Button (`only_wordpressdotcom`), and list (`include_wordpressdotcom`):

![image](https://user-images.githubusercontent.com/96308/69355206-9bbcbd00-0cab-11ea-90ad-e035031327b5.png)

:man_facepalming:

sirreal commented:
> Enter something into the site title field, then delete again. DomainPickerButton continues to show the free domain for that site title.

This was intended, it seems strange to show the interactive element, then hide it again. It can be removed.

sirreal commented:
> I seem to get different free WP.com subdomain suggestions for `Roasting Beans` in the Domain Picker Button (`only_wordpressdotcom`), and list (`include_wordpressdotcom`):

I also noticed that results were very different. `only_wordpressdotcom` would actually fail for many searches. I dropped its usage and added a note to avoid it.

ockham commented on client/landing/gutenboarding/components/header/index.tsx:
```diff
@@ -31,7 +34,41 @@ const Header: FunctionComponent< Props > = ( {
 	toggleSidebarShortcut,
 } ) => {
-	const { siteTitle, siteVertical } = useSelect( select => select( STORE_KEY ).getState() );
+	const [ domainText, setDomainText ] = useState< string | null >( null );
```
Do we need the `| null`? Can we use empty string instead? (To keep the type more... homogenous?)

ockham replied:
```suggestion
	const [ domainText, setDomainText ] = useState< string >( '' );
```

ockham replied:
Actually :sweat_smile: 

```suggestion
	const [ domainText, setDomainText ] = useState< DomainName >( '' );
```
